src/api/messageprocessors/UpdateListingItemMessageProcessor.ts

```diff
@@ -52,30 +54,47 @@ export class UpdateListingItemMessageProcessor implements MessageProcessorInterf
         return await this.listingItemService.update(listingItemTobeUpdate.id, listingItem as ListingItemUpdateRequest);
     }
 
-    private async createCategoryIfNotExist(category: string[]): Promise<number> {
+    private async createCategoryIfNotExist(categoryArray: string[]): Promise<ItemCategory> {
         const rootCategoryWithRelated: any = await this.itemCategoryService.findRoot();
-        const itemCategoryOutPut: any = await this.itemCategoryFactory.getModel(category, rootCategoryWithRelated);
-        const itemCategory = itemCategoryOutPut.createdCategories;
-        const lastCreatedIndex = itemCategoryOutPut.lastCheckIndex;
-        let itemCategoryId: number;
-        let parentItemCategoryId;
-        if (lastCreatedIndex === category.length) {
-            // all category is exist
-            itemCategoryId = itemCategory[lastCreatedIndex].id;
-        } else {
-            // get and create category
-            let newCat;
-            parentItemCategoryId = itemCategory[lastCreatedIndex].id;
-            for (let i = lastCreatedIndex + 1; i < category.length; i++) {
-                newCat = await this.itemCategoryService.create({
-                    name: category[i],
-                    parent_item_category_id: parentItemCategoryId
-                } as ItemCategoryCreateRequest);
-                parentItemCategoryId = newCat.id;
+        let parentItemCategoryId = 0;
+        let returnCategory;
+        for (const category of categoryArray) { // [cat0, cat1, cat2, cat3, cat4]
+            const catExist = await this.checkCategory(rootCategoryWithRelated, category);
+            let categoryExist;
+            if (!catExist) {
+                // not found
+                const categoryCreateReqMessage = await this.itemCategoryFactory.getModel({
+                    name: category,
+                    parentItemCategoryId
+                } as ItemCategoryMessage);
+                // check with parenID and name
+                categoryExist = await this.itemCategoryService.isCategoryExists(
+                    categoryCreateReqMessage.name,
+                    returnCategory // as parentCategory
+                );
+                if (categoryExist === null) {
+                    // create and return Id
+                    categoryExist = await this.itemCategoryService.create(categoryCreateReqMessage);
+                }
+            } else {
+                categoryExist = await this.itemCategoryService.findOneByKey(category);
             }
-            itemCategoryId = newCat.id;
+            parentItemCategoryId = categoryExist.id;
+            // parentCategory = categoryExist;
+            returnCategory = categoryExist;
+        }
+        return returnCategory as ItemCategory;
+    }
+
+    private async checkCategory(categories: ItemCategory, value: string): Promise<any> {
+        if (categories['key'] === value) { // check cat_ROOT
```

Comment:
rename, so that it's clear what's being checked just by looking at the function name.